include/oneapi/dpl/internal/reduce_by_segment_impl.h

```diff
@@ -160,25 +161,17 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
     typedef typename ::std::iterator_traits<InputIterator2>::value_type ValueType;
     typedef uint64_t CountType;
     typedef typename ::std::decay<Policy>::type policy_type;
+    typedef typename ::std::iterator_traits<InputIterator1>::value_type DerefValueType;
 
     namespace __bknd = __par_backend_hetero;
 
-    ValueType initial_value;
-    {
-        auto first2_acc = internal::get_access<sycl::access::mode::read>(policy, first2);
-        initial_value = first2_acc[0];
-    }
-
     const auto n = ::std::distance(first1, last1);
     if (n <= 0)
         return ::std::make_pair(result1, result2);
-    else if (n == 1)
+    if (n == 1)
     {
-        auto result1_acc = internal::get_access<sycl::access::mode::write>(policy, result1);
-        auto result2_acc = internal::get_access<sycl::access::mode::write>(policy, result2);
-        auto first1_acc = internal::get_access<sycl::access::mode::read>(policy, first1);
-        result1_acc[0] = first1_acc[0];
-        result2_acc[0] = initial_value;
+        internal::set_data_at(policy, result1, 0, internal::get_data_at<DerefValueType>(policy, first1, 0));
+        internal::set_data_at(policy, result2, 0, internal::get_data_at<ValueType>(policy, first2, 0));
```

Comment:
Please see my comment above regarding overhead "device-to-host-to-device" copying.

Reply:
Fixed